File: utils/passwordValidator.ts
/**
 * Password validation utility
 * Requirements:
 * - Minimum 8 characters
 * - At least 1 uppercase letter
 * - At least 1 lowercase letter
 * - At least 1 digit
 */

export const passwordRequirements = {
  minLength: 8,
  hasUppercase: false,
  hasLowercase: false,
  hasDigit: false,
};

export interface PasswordValidation {
  isValid: boolean;
  minLength: boolean;
  hasUppercase: boolean;
  hasLowercase: boolean;
  hasDigit: boolean;
  errors: string[];
}

export const validatePassword = (password: string): PasswordValidation => {
  const errors: string[] = [];
  
  const minLength = password.length >= 8;
  if (!minLength) errors.push('Minimum 8 caractères');

  const hasUppercase = /[A-Z]/.test(password);
  if (!hasUppercase) errors.push('Au moins 1 lettre majuscule');

  const hasLowercase = /[a-z]/.test(password);
  if (!hasLowercase) errors.push('Au moins 1 lettre minuscule');

  const hasDigit = /\d/.test(password);
  if (!hasDigit) errors.push('Au moins 1 chiffre');

  return {
    isValid: minLength && hasUppercase && hasLowercase && hasDigit,
    minLength,
    hasUppercase,
    hasLowercase,
    hasDigit,
    errors,
  };
};

export const getPasswordStrength = (validation: PasswordValidation): 'faible' | 'moyen' | 'fort' => {
  if (!validation.isValid) return 'faible';
  if (validation.hasUppercase && validation.hasLowercase && validation.hasDigit) return 'fort';
  return 'moyen';
};

File: backend/app/utils/password_validator.py
"""
Password validation utility for backend.
Ensures all passwords meet security requirements:
- Minimum 8 characters
- At least 1 uppercase letter (A-Z)
- At least 1 lowercase letter (a-z)
- At least 1 digit (0-9)
"""

from pydantic import BaseModel
import re


class PasswordValidation(BaseModel):
    is_valid: bool
    min_length: bool
    has_uppercase: bool
    has_lowercase: bool
    has_digit: bool
    errors: list[str] = []


def validate_password(password: str) -> PasswordValidation:
    """
    Validate password meets all security requirements.
    
    Args:
        password: Password to validate
        
    Returns:
        PasswordValidation object with validation results
    """
    errors = []
    
    # Check minimum length
    min_length = len(password) >= 8
    if not min_length:
        errors.append("كلمة المرور يجب أن تكون 8 أحرف على الأقل")
    
    # Check for uppercase letters
    has_uppercase = bool(re.search(r'[A-Z]', password))
    if not has_uppercase:
        errors.append("كلمة المرور يجب أن تحتوي على حرف كبير واحد على الأقل")
    
    # Check for lowercase letters
    has_lowercase = bool(re.search(r'[a-z]', password))
    if not has_lowercase:
        errors.append("كلمة المرور يجب أن تحتوي على حرف صغير واحد على الأقل")
    
    # Check for digits
    has_digit = bool(re.search(r'\d', password))
    if not has_digit:
        errors.append("كلمة المرور يجب أن تحتوي على رقم واحد على الأقل")
    
    is_valid = all([min_length, has_uppercase, has_lowercase, has_digit])
    
    return PasswordValidation(
        is_valid=is_valid,
        min_length=min_length,
        has_uppercase=has_uppercase,
        has_lowercase=has_lowercase,
        has_digit=has_digit,
        errors=errors
    )
